File: herencia-dto/SergioDeIscar-herencia-java/SergioDeIscar-herencia-java/src/main/java/org/example/services/storage/persona/PersonaFileJson.java
package org.example.services.storage.persona;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import org.example.config.AppConfig;
import org.example.dto.PersonaDto;
import org.example.dto.PersonasDto;
import org.example.models.Persona;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.util.List;

public class PersonaFileJson implements PersonaStorageService{
    // Singleton
    private static PersonaFileJson instance = null;
    public static PersonaFileJson getInstance() {
        if (instance == null) {
            instance = new PersonaFileJson();
        }
        return instance;
    }

    private final String localFile = AppConfig.getInstance().getAppData() + File.separator + "personas.json";
    private final Moshi moshi = new Moshi.Builder().build();
    @Override
    public List<Persona> saveAll(List<Persona> elements) {
        File file = new File(localFile);
        if (file.exists() && !file.canWrite() ) return null;
        JsonAdapter<PersonasDto> jsonAdapter = moshi.adapter(PersonasDto.class);
        try(PrintWriter pw = new PrintWriter(new FileWriter(file))) {
            pw.println(
                    jsonAdapter.indent("\t").toJson(
                        new PersonasDto(
                            elements.stream().map(Persona::toDto).toList()
                        )
                    )
            );
        }catch (IOException e){
            System.out.println("Error al escribir en el fichero");
        }
        return elements;
    }

    @Override
    public List<Persona> loadAll() {
        File file = new File(localFile);
        if(!file.exists() || !file.canRead()) return null;
        JsonAdapter<PersonasDto> jsonAdapter = moshi.adapter(PersonasDto.class);
        try {
            String json = Files.readString(file.toPath());
            PersonasDto dto = jsonAdapter.fromJson(json);
            return dto.getPersonas().stream().map(PersonaDto::toClass).toList();
        }catch (IOException e){
            System.out.println("Error al leer en el fichero");
        }
        return null;
    }
}

File: herencia-dto/SergioDeIscar-herencia-kotlin/SergioDeIscar-herencia-kotlin/src/main/kotlin/service/storage/persona/PersonaFileJson.kt
package service.storage.persona

import com.squareup.moshi.Moshi
import com.squareup.moshi.adapter
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import config.AppConfig
import dto.PersonaDto
import mappers.toClass
import mappers.toDto
import models.Persona
import mu.KotlinLogging
import java.io.File
import java.io.IOException

private val logger = KotlinLogging.logger {}

@ExperimentalStdlibApi
object PersonaFileJson: PersonaStorageService {
    private val localFile = "${AppConfig.APP_DATA}${File.separator}persona.json"
    private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()

    override fun saveAll(elements: List<Persona>): List<Persona> {
        logger.debug { "PersonaFileJson ->\tsaveAll" }
        val file = File(localFile)
        if (file.exists() && !file.canWrite() ) throw IOException("No se puede escribir en el archivo")

        val jsonAdapter = moshi.adapter<List<PersonaDto>>()
        file.writeText(
            jsonAdapter.indent("\t").toJson(
                elements.map { it.toDto() }
            )
        )
        return elements
    }

    override fun loadAll(): List<Persona> {
        logger.debug { "PersonaFileJson ->\tloadAll" }
        val file = File(localFile)
        if(!file.exists() || !file.canRead()) throw IOException("No se puede leer el archivo")

        val jsonAdapter = moshi.adapter<List<PersonaDto>>()
        return jsonAdapter.fromJson(file.readText())?.map { it.toClass() } ?: emptyList()
    }
}